app/vmalert/main.go

```diff
@@ -44,15 +63,69 @@ func main() {
 	if err := httpserver.Stop(*httpListenAddr); err != nil {
 		logger.Fatalf("cannot stop the webservice: %s", err)
 	}
+	cancel()
 	w.stop()
 }
 
 type watchdog struct {
-	storage *datasource.VMStorage
+	stg *datasource.VMStorage
+	p   provider.AlertProvider
+}
+
+func (w *watchdog) run(ctx context.Context, a config.Group, evaluationInterval time.Duration) {
+	t := time.NewTicker(evaluationInterval)
+	var metrics []datasource.Metric
+	var err error
+	var alerts []provider.Alert
+	defer t.Stop()
+	for {
+		select {
+		case <-t.C:
```

Comment:
sequential evaluation of rule-set may take more then evaluationInterval. Should we consider independent goroutines for each rule or rule-group?

Reply:
vmalert runs single go-routine for group